core/server/common/src/main/java/alluxio/master/journal/raft/RaftJournalSystem.java

```diff
@@ -327,6 +334,58 @@ public synchronized void losePrimacy() {
     LOG.info("Raft server successfully restarted");
   }
 
+  @Override
+  public synchronized void checkpoint() throws IOException {
+    mSnapshotAllowed.set(true);
+    CopycatClient client = createAndConnectClient();
+    try {
+      long start = System.currentTimeMillis();
+      LOG.info("Submitting empty journal entry to trigger snapshot");
+      // If snapshot requirements are fulfilled, a snapshot will be triggered
+      // after sending an empty journal entry to Copycat
+      CompletableFuture<Void> future = client.submit(new JournalEntryCommand(
+          JournalEntry.getDefaultInstance()));
+      try {
+        future.get(1, TimeUnit.MINUTES);
+      } catch (TimeoutException | ExecutionException e) {
+        LOG.info("Exception submitting entry to trigger snapshot: {}", e.toString());
+        throw new IOException("Exception submitting entry to trigger snapshot", e);
+      } catch (InterruptedException e) {
+        LOG.info("Interrupted when submitting entry to trigger snapshot: {}", e.toString());
+        Thread.currentThread().interrupt();
+        throw new CancelledException("Interrupted when submitting entry to trigger snapshot", e);
+      }
+      if (mStateMachine.getLastSnapshotStartTime() < start) {
+        throw new IOException("Do not fulfill Copycat snapshot requirements. "
+            + "No snapshot is triggered");
+      }
+      if (mStateMachine.isSnapshotting()) {
+        waitForSnapshotting(mStateMachine);
+      }
+      LOG.info("Snapshotted in raft journal system");
+    } finally {
+      mSnapshotAllowed.set(false);
+    }
+  }
+
+  /**
+   * Waits for snapshotting to finish.
+   *
+   * @param stateMachine the journal state machine
+   */
+  private void waitForSnapshotting(JournalStateMachine stateMachine) throws IOException {
+    try {
+      CommonUtils.waitFor("snapshotting to finish", () -> !stateMachine.isSnapshotting(),
+          WaitForOptions.defaults().setTimeoutMs(20 * Constants.MINUTE_MS));
```

Comment:
unrelated to this PR, but can we make this configurable and increase the default to 2 hours? With 1 billion files this might time out due to the large scale

Reply:
Add a property key to control how long we wait for the snapshotting to finish.